src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java

```diff
@@ -33,39 +38,88 @@ public void run() {
+      FlowConnectionProvider connectionProvider)
+      throws ExecutionException, InterruptedException, IOException {
+    while (!flowSteps.isEmpty()) {
+      PacketStep step = flowSteps.poll();
+      if (step instanceof ParallelPacketStep) {
+        processParallelSteps((ParallelPacketStep) step, previousPort, connectionProvider);
+        continue;
+      }
+      if (step.getPort() != null && step.getPort() != previousPort) {
+        LOG.info("changing to connections on port {}", step.getPort());
+        previousPort = step.getPort();
+      }
+      FlowConnection flowConnection = connectionProvider.get(previousPort);
+      step.process(flowConnection);
+    }
+  }
+
+  private static void processParallelSteps(ParallelPacketStep step, int previousPort,
+      FlowConnectionProvider connectionProvider) {
+    List<List<PacketStep>> parallelSteps = step.getParallelSteps();
+    ExecutorService parallelStepsExecutor = Executors.newFixedThreadPool(parallelSteps.size());
+    for (List<PacketStep> parallelStep : parallelSteps) {
+      Queue<PacketStep> steps = new LinkedList<>(parallelStep);
+      parallelStepsExecutor.submit(() -> {
+        try {
+          processSteps(steps, previousPort, connectionProvider);
+        } catch (ExecutionException | InterruptedException | IOException e) {
+          handleProcessStepExceptions(e);
+        }
+      });
+    }
+    joinParallelExecutions(parallelStepsExecutor);
+  }
+
+  private static void handleProcessStepExceptions(Exception e) {
+    if (e.getClass().equals(IOException.class)) {
       if (e.getMessage().contains("Socket is closed")) {
         LOG.trace("Received expected exception when server socket has been closed", e);
       } else {
         LOG.error("Problem while processing requests from client. Closing connection.", e);
       }
-    } catch (InterruptedException e) {
+    } else if (e.getClass().equals(InterruptedException.class)) {
       LOG.trace("The thread has been interrupted", e);
       Thread.currentThread().interrupt();
-    } catch (ExecutionException e) {
+    } else if (e.getClass().equals(ExecutionException.class)) {
       LOG.error("Problem while waiting for socket to be created", e);
-    } finally {
+    } else if (e.getClass().equals(ConnectionClosedException.class)) {
+      LOG.info("Connection closed by client while waiting for client packet");
+      ConnectionClosedException ex = (ConnectionClosedException) e;
+      if (ex.getDiscardedPacket().getBytes().length > 0) {
+        LOG.debug("Discarding client packet {}", ex.getDiscardedPacket(), e);
+      }
+    }
+  }
+
+  private static void joinParallelExecutions(ExecutorService parallelStepsExecutor) {
+    parallelStepsExecutor.shutdown();
+    while (true) {
       try {
-        closeFlowConnections();
-      } catch (IOException e) {
-        LOG.error("Problem while releasing sockets", e);
+        if (parallelStepsExecutor.awaitTermination(PARALLEL_THREADS_TIMEOUT,
```

Comment:
we are imposing a timeout regardless of the number of steps to the total execution time of sub flows. This sounds fragile and require constant tuning of the timeout for users.

Reply:
Maybe, to avoid leaving connections open, and not just for parallel execution, we should put some timeout in general of flows activity? Something like: all connections of a flow will be closed after X time of no received packet? Another thing you can do is not put a timeout, and in this case, collect all Futures from submit, iterate over them doing a get, and then shutdown the thread pool.

Reply:
Any idea to face the issue?
 Nowadays when client does not complete the whole flow we are up listening indefinitely. How is that different from this scenario?

Reply:
I took the second idea. The one using Futures get. Since I believe it keeps aligned with previous Wiresham versions idea. The idea of listening indefinitely until a connection is made.